### PR TITLE
Created DataAccess for search and related classes.

### PR DESCRIPTION
**Pull request by EquilibriaW**

### CODE REVIEW

**Comment by uoftm** [comment]
Can you format your code here

File: src/entity/ChatMessage.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import java.util.List;
+
+public interface ChatMessage {
+    public String getTime();
+    public String getMessage();
+    public List<User> getUsers();
+
+
+
+}
```
Comment (alexrosen45) [inline src/entity/ChatMessage.java:12]: Please format your code before committing using `mvn com.spotify.fmt:fmt-maven-plugin:format`

File: src/entity/SearchRequest.java [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package entity;
+
+public class SearchRequest {}
```
Comment (alexrosen45) [inline src/entity/SearchRequest.java:3]: Why is this here already without Elasticsearch functionality?

File: src/entity/ChatMessage.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package entity;
+
+import java.util.List;
+
+public interface ChatMessage {
+    public String getTime();
+    public String getMessage();
+    public List<User> getUsers();
```
Comment (alexrosen45) [inline src/entity/ChatMessage.java:8]: It's good to have a list of `User`s here, but you should add a `uid` attribute to the `User` class if you want to be able to use this properly.

File: src/data_access/SearchDataAccessObject.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      }
+    } catch (Exception e) {
+      searchRequest.changeQueryResponse("sorry, didn't find anything!");
+    }
+  }
+
+  @Override
+  public void saveData(SearchChatMessage message) {
+    // Replace these placeholders with your actual Elasticsearch URL and API key
+    String esUrl = "your_elasticsearch_url";
```
Comment (uoftm) [inline src/data_access/SearchDataAccessObject.java:47]: Please extract esUrl and apiKey into https://github.com/uoftm/csc207project/blob/main/src/data_access/Constants.java

File: src/data_access/SearchDataAccessObject.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        {"name": "1984", "author": "George Orwell", "release_date": "1985-06-01", "page_count": 328, "_extract_binary_content": true, "_reduce_whitespace": true, "_run_ml_inference": false}
+        { "index" : { "_index" : "search-chats" } }
+        {"name": "Fahrenheit 451", "author": "Ray Bradbury", "release_date": "1953-10-15", "page_count": 227, "_extract_binary_content": true, "_reduce_whitespace": true, "_run_ml_inference": false}
+        { "index" : { "_index" : "search-chats" } }
+        {"name": "Brave New World", "author": "Aldous Huxley", "release_date": "1932-06-01", "page_count": 268, "_extract_binary_content": true, "_reduce_whitespace": true, "_run_ml_inference": false}
+        { "index" : { "_index" : "search-chats" } }
+        {"name": "The Handmaid's Tale", "author": "Margaret Atwood", "release_date": "1985-06-01", "page_count": 311, "_extract_binary_content": true, "_reduce_whitespace": true, "_run_ml_inference": false}
+        """;
+
+    OkHttpClient client = new OkHttpClient();
```
Comment (uoftm) [inline src/data_access/SearchDataAccessObject.java:67]: Consider taking this client as a parameter in the constructor to this class (and making client a private instance variable)

File: src/data_access/SearchDataAccessObject.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    String esUrl = "";
+    String apiKey = "";
+
+    OkHttpClient client = new OkHttpClient();
+
+    // Define the Elasticsearch index to query
+    String index = "search-chats";
+
+    // Define your Elasticsearch query (in this case, a match query for the "name" field)
+    String query = "{ \"query\": { \"match\": { \"name\": \"1984\" } } }";
```
Comment (uoftm) [inline src/data_access/SearchDataAccessObject.java:23]: maybe use a json stringifier (see https://github.com/uoftm/csc207project/blob/main/src/data_access/FirebaseMessageDataAccessObject.java#L24C26-L24C85) to build a json object from the searchRequest and put it here

File: pom.xml [hunk 7]
```diff
@@ -20,6 +20,12 @@
             <artifactId>okhttp</artifactId>
             <version>4.11.0</version>
         </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
```
Comment (uoftm) [inline pom.xml:25]: Yay, testing! Any chance you can add back your tests?

**Review by uoftm — COMMENTED** [review]
This has a bunch of awesome stuff! Testing, new apis (elasticsearch), and a new interactor.

File: src/entity/ChatMessage.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package entity;
```
Comment (uoftm) [inline src/entity/ChatMessage.java:1]: Oops, we now have a Mesage.java entity, sorry for leaving this hanging for so long!

File: src/entity/SearchChatMessage.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package entity;
+
+import java.time.LocalDateTime;
+
+public class SearchChatMessage implements ChatMessage {
+
+  private final LocalDateTime messageTime;
```
Comment (uoftm) [inline src/entity/SearchChatMessage.java:7]: I believe we should use Instant here instead of LocalDateTime because this is what is stored on the server (and use LocalDateTime for messages that are displayed to the user). Does that seem right?

File: src/entity/SearchRequest.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
+package entity;
+
+public class SearchRequest {
+
+  private final String queryRequest;
+
+  private String queryResponse;
```
Comment (uoftm) [inline src/entity/SearchRequest.java:7]: The query response is a structured JSON object right? Can I see a sample response, and then I can help construct a better data type here?

Comment (uoftm) [reply src/entity/SearchRequest.java:7]: Also, for clean architecture you'll need to split this into two files (request and response)

File: tests/SearchTest.java [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import static org.junit.Assert.*;
+
+
+
+public class SearchTest {
+
```
Comment (uoftm) [inline tests/SearchTest.java:6]: Looking forward to seeing the test here ;)

File: src/entity/ChatMessage.java (hunk 8 above)
Comment (EquilibriaW) [reply src/entity/ChatMessage.java:1]: Ok, resolved by making chatmessage child of message.

File: src/entity/SearchChatMessage.java (hunk 9 above)
Comment (EquilibriaW) [reply src/entity/SearchChatMessage.java:7]: Sure

File: src/data_access/SearchDataAccessObject.java (hunk 4 above)
Comment (EquilibriaW) [reply src/data_access/SearchDataAccessObject.java:47]: Ok, done

File: tests/SearchTest.java (hunk 11 above)
Comment (EquilibriaW) [reply tests/SearchTest.java:21]: Added first test

File: pom.xml (hunk 7 above)
Comment (EquilibriaW) [reply pom.xml:25]: yes, done

File: src/data_access/SearchDataAccessObject.java (hunk 6 above)
Comment (EquilibriaW) [reply src/data_access/SearchDataAccessObject.java]: Done

File: src/entity/SearchRequest.java (hunk 10 above)
Comment (EquilibriaW) [reply src/entity/SearchRequest.java]: yeah, added new classes, and implementation of the data access object.

File: src/data_access/SearchDataAccessObject.java (hunk 5 above)
Comment (EquilibriaW) [reply src/data_access/SearchDataAccessObject.java:114]: what does this mean?

File: csc207project.iml [hunk 12]
```diff
@@ -3,6 +3,8 @@
   <component name="AdditionalModuleElements">
     <content url="file://$MODULE_DIR$" dumb="true">
       <sourceFolder url="file://$MODULE_DIR$/src" isTestSource="false" />
+      <sourceFolder url="file://$MODULE_DIR$/test" isTestSource="true" />
+      <sourceFolder url="file://$MODULE_DIR$/tests" isTestSource="true" />
```
Comment (uoftm) [inline csc207project.iml:7]: One of these lines is intentional, one should be deleted

File: src/data_access/SearchDataAccessObject.java (hunk 5 above)
Comment (uoftm) [reply src/data_access/SearchDataAccessObject.java]: Here's what I did in FirebaseMessageDataAccessObject, and I would recommend imitating that here (because using only a single okhttpclient is faster)

```java

  private final OkHttpClient client;

  public FirebaseMessageDataAccessObject(OkHttpClient client) {
    this.client = client;
  }
```

Comment (EquilibriaW) [reply src/data_access/SearchDataAccessObject.java]: Ok, modified this.

File: src/entity/ChatMessage.java (hunk 3 above)
Comment (EquilibriaW) [reply src/entity/ChatMessage.java]: removed list of users, made it purely about roomID.

**Review by uoftm — APPROVED** [review]
Awesome work!
